executor/builder.go

```diff
@@ -2228,41 +2215,12 @@ func (b *executorBuilder) updateForUpdateTSIfNeeded(selectPlan plannercore.Physi
 	// The Repeatable Read transaction use Read Committed level to read data for writing (insert, update, delete, select for update),
 	// We should always update/refresh the for-update-ts no matter the isolation level is RR or RC.
 	if b.ctx.GetSessionVars().IsPessimisticReadConsistency() {
-		return b.refreshForUpdateTSForRC()
+		_, err = sessiontxn.GetTxnManager(b.ctx).GetStmtForUpdateTS()
+		return err
 	}
 	return UpdateForUpdateTS(b.ctx, 0)
 }
 
-// refreshForUpdateTSForRC is used to refresh the for-update-ts for reading data at read consistency level in pessimistic transaction.
-// It could use the cached tso from the statement future to avoid get tso many times.
-func (b *executorBuilder) refreshForUpdateTSForRC() error {
```

Comment:
Ditto, would the `updateForUpdateTSIfNeeded` be refactored too in the future?

Reply:
Yes, I want to remove it in the future. But we should still keep it when all codes are moved to provider. Because for `Insert`  the `GetStmtForUpdateTS` should be called to update the snapshot's ts for the later update operations. However after refactor, `updateForUpdateTSIfNeeded` will be more simple and just calling`TxnManager.GetStmtForUpdateTS` is enough.

Maybe another way is to add a new method `TxnForWrite` to `TxnManager` and the `InsertExec` should use it instead of `sessionctx.Txn`